### PR TITLE
Make the multiplayer faction accessible via Lua and simplified Hornshark Island

### PR DESCRIPTION
**Pull request by spixi**
Successor of #658

I hope, I finally have it.

The attribute _faction_ was already used to set the default faction in a scenario. I think, it would be save to override it, however. So here it is.

### CODE REVIEW

**Comment by gfgtdf** [comment]
You shouldn't make a new pr each time you change something this make it very hard to follow the discussion log.

Your commit adds an entry install_manifest.txt in the .gitignore, not sure if this is intended, if it is then you shodul put that in a  different  commit so that becasue it seems unrelated to the other changes.

**Comment by spixi** [comment]
Hello @gfgtdf,
I really don't know how I can delete commits, which I have already made to my github fork spixi/wesnoth. CVS is so much easier, because you have a history per file, but Git has only a history for the whole repository, which makes deployment a misery. I think the addition in .gitignore is okay, because this is a file created by CMake, which is not needed in the Git repository.

**Comment by spixi** [comment]
I learned how to work with branches now. Should I create another PR or do you think, you could merge this? I also noted, that I added the note to the wrong version in the changelog.

**Comment by gfgtdf** [comment]
The problem here is that the commit history is bad, and this is exactly the thing that cannot be fixed after the pr is mergd.

> Should I create another PR 

Hmm no, just to it in the branch of this pr and then squash the commits.

**Comment by spixi** [comment]
This PR only has one commit. I unfortunately don't know how to delete commits. Rebase and squash produces additional commits. I should never try to do changes in the master branch anymore ... Maybe I can fix it, but not today.

**Comment by gfgtdf** [comment]
Hmm what commands do you use for rebase & sqauash ?

**Comment by spixi** [comment]
git rebase <commit id>

**Comment by gfgtdf** [comment]
no, i meant more exactly. like for example 'first ``git rebase -i HEAD~3` then change a line to squash, then `git push -f`'

**Comment by spixi** [comment]
So I ran `git rebase -I 68e5e7d`
Then vi opened. I changed to the insert mode with `i` and I replaced the second `pick` with `fixup`.
I saved and quit with `<ESC>:wq<Enter>`.
Then I excecuted `git push`, inserted my user name and password.

I get the following output:
```
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'https://github.com/spixi/wesnoth.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

**Comment by spixi** [comment]
So, the next try with `git pull` and `git push --force`, but it will still generate more commits instead of squashing commits. What do I do wrong?

**Comment by spixi** [comment]
Oh, apparently, `git push --force` actually worked :)

**Comment by gfgtdf** [comment]
OK this commit has still some issues, inparticular it changes `.gitignore` and `lua_interpreter.cpp
` which both seem to be quiet unrelated to the rest of that commit. please split those parts form that commit into other commits. This can for example be done by making a commit B that undoes those changes, then make a revert commit of B, and then squash commit B into the first commit with rebase.

**Comment by spixi** [comment]
Fixed it.

**Comment by gfgtdf** [comment]
Hmm the travis tests on this branch fail. Since they don't fail when the key was still named 'faction_id' i suggested to switch it back to 'faction_id' unless you know how to fix those travis tests.

**Comment by spixi** [comment]
A unit test fails.

It seems that there is an undocumented feature of a "Custom" faction. If the faction is set to "Custom" in the `[side]` tag of the multiplayer campaign, you have the possibility to pick any possible leader in the era. However, you get no recruit list.

I don't know, whether this is a relict from a former development, but I think this is very weird. Is that "Custom" faction still needed?

**Comment by spixi** [comment]
I found that faction=Custom is actually needed here:

**Comment by spixi** [comment]
I think the unit test can be fixed by setting  faction_lock=true. I will try it.

**Comment by spixi** [comment]
See also commit 9dae74cd5de94319b038674837937308abcbb9d4 where some of these tests were disabled.

**Comment by spixi** [comment]
@thunderstruck Can you have a look upon these failing unit tests? Are they still up-to-date?

**Comment by gfgtdf** [comment]
The 'Custom' faction is defeined here: https://github.com/wesnoth/wesnoth/blob/1cc3711996923636e5e979b00dbd9f7e5dbd24d8/data/_main.cfg#L20 

The reason why LoW uses it is becasue t doesn'T want recuitlist from a random factiosn added to the side.

I don't think thunderstruck is active anymore, so even if he gets notified he mostlikeley won't remember anymore.

**Comment by spixi** [comment]
The following test case is failing:

```
// Carried over recruits. 
side.clear(); 
side["previous_recruits"] = "Elvish Archer"; 
side_engine.reset(create_mp_side_engine(side, connect_engine.get())); 
BOOST_CHECK( side_engine->flg().choosable_factions().size() >  1 ); 
BOOST_CHECK_EQUAL( side_engine->flg().current_faction()["id"], "Custom" ); 
BOOST_CHECK_EQUAL( side_engine->new_config()["previous_recruits"], "Elvish Archer" ); 
```
I actually don't know why `side_engine->flg().current_faction()["id"]` even should have the value `"Custom"` although we cleared the side before. I think we are testing against an undefined variable. So adding a `side["faction"] = "Custom";` under `side.clear();` should fix the unit test. The other possibility is to change the `clear()` method, but I don't know which side effects this would case.

**Comment by gfgtdf** [comment]
Hmm this is beasue whnever you have already a recuitlist  defined in the scenario [side], then faction= should default to "Custom" to not change the recuitlist.

**Comment by spixi** [comment]
I played a little bit with Legend of Wesmere's cfg files and found out, that everything is fine.

But I found the issue now. In `src/game_initialization/connect_engine.cpp`, method `side_engine::reset` the current faction is set to 0 unless the game is a saved game.

I don't know what is wrong with the unit test though. Everything works fine.

```
diff --git a/src/game_initialization/connect_engine.cpp b/src/game_initialization/connect_engine.cpp
index 41064e5..ba346b4 100644
--- a/src/game_initialization/connect_engine.cpp
+++ b/src/game_initialization/connect_engine.cpp
@@ -987,6 +987,8 @@ config side_engine::new_config() const
        if (!parent_.params_.saved_game) {
                // Merge the faction data to res.
                config faction = flg_.current_faction();
+               std::cout << "Old: " << res["faction"].str() << res["faction_name"].str() << std::endl; //DEBUG
+               std::cout << "New: " << faction["name"].str() << faction["id"].str() << std::endl << std::endl; //DEBUG
                res["faction_name"] = faction["name"];
                res["faction"] = faction["id"];
                faction.remove_attributes("id", "name", "image", "gender", "type");
diff --git a/src/game_initialization/flg_manager.cpp b/src/game_initialization/flg_manager.cpp
index c0bad05..2b0cba1 100644
--- a/src/game_initialization/flg_manager.cpp
+++ b/src/game_initialization/flg_manager.cpp
@@ -99,6 +99,7 @@ flg_manager::flg_manager(const std::vector<const config*>& era_factions,
        update_available_factions();
 
        set_current_faction(0);
+    std::cout << "current_faction: " << (*current_faction_)["id"].str() << std::endl; //DEBUG
 }
 
 flg_manager::~flg_manager()
@@ -555,6 +556,7 @@ int flg_manager::find_suitable_faction() const
                }
                ++index;
        }
+
        return res;
 }

```

**Comment by gfgtdf** [comment]
Hmm i don'T rally see how side_engine::reset is related to this issue since afaik 'side_engine::reset' is onyl called when a player leaves.

**Comment by gfgtdf** [comment]
Hmm i just tested an it seems liek the enemies in 4p a new land now recuits units from a random faction, not ure if this is caused by this pr though

**Comment by gfgtdf** [comment]
i made a comit to fix the boosts test: https://github.com/wesnoth/wesnoth/commit/e9d7efbd1ea7bd3a892c19ee5c499dfc602f33c1

assuming that you tested all other things i'll merge this.

**Comment by spixi** [comment]
Do you know who the maintainer of Hornshark Island is? I noticed that there is no specific handler for Khalifate. There is a fallback which spawns some monsters. I don't know whether this is balanced with Khalifate, however.

**Comment by gfgtdf** [comment]
I dont' think that there is currently a maintainer for hornshark island. Its quiet posible that it was written before that faction was added, note that the fallback case is also important for the case when the scenario is played with a completeley different era.

**Comment by spixi** [comment]
As long there are no complaints about it, I think it's okay. Having some monsters in the "mythical" Khalifate faction is even very cool.

**Comment by CelticMinstrel** [comment]
The Khalifate faction doesn't seem mythical to me, honestly.

**Comment by spixi** [comment]
Unfortunately the Khalifate have no Level 0 units like Peasant or Ruffian (Falcon does not count).

Do you think, 1 loyal Young Ogre, 1 loyal Jundi, 1 loyal Naffat  and 1 loyal Khaiyal would be balanced?

**Comment by CelticMinstrel** [comment]
I haven't played Hornshark Island. Why are level 0 units relevant and why does the falcon not count? (Khalifate aren't the only faction to lack level 0 units, anyway.)

**Comment by spixi** [comment]
Loyalists, Knalgan and Undead get some Level 0 units for starting. Northerners get Orcish Leader to support Goblins. Only Rebels and Drakes get regular units. Drakes are severely handicapped with a slow and dim Fighter, because of their map advantage.

I did not count Falcon and Vampire Bat, because they would be too strong as starting units due to their movement type.

Oh, I just saw that the fallback unit set is compiled with Khalifate in mind: https://forums.wesnoth.org/viewtopic.php?p=518542&sid=73489ef47cba8b3a39ac219353f64613#p518542

**Comment by spixi** [comment]
In this forum thread Doc Paterson states, that he was working on a unit set for Khalifate. Not sure, if he already has one which uses Khalifate units.
